pkg/controllers/csi/provisioner/cleanup/run.go

```diff
@@ -0,0 +1,166 @@
+package cleanup
+
+import (
+	"context"
+	"os"
+
+	dtcsi "github.com/Dynatrace/dynatrace-operator/pkg/controllers/csi"
+	"github.com/Dynatrace/dynatrace-operator/pkg/controllers/csi/metadata"
+	"github.com/spf13/afero"
+	"k8s.io/mount-utils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Cleaner struct {
+	fs        afero.Afero
+	apiReader client.Reader
+	mounter   mount.Interface
+	path      metadata.PathResolver
+}
+
+// fsState collects all the "top-level" folders we care about and categorizes them
+type fsState struct {
+	// deprecatedDks are the dynakube-dirs under /data that have a /run directory, which used to contain the app-mounts, these directories are named after the tenantUUID
+	deprecatedDks []string
+	// binDks are dynakube-dirs that have a "latest-symlink" pointing at a codemodule binary
+	binDks []string
+	// hostDks are dynakube-dirs that contain the folder that was mounted to Host OneAgents
+	hostDks []string
+}
+
+func New(fs afero.Afero, apiReader client.Reader, path metadata.PathResolver) *Cleaner {
+	return &Cleaner{
+		fs:        fs,
+		apiReader: apiReader,
+		path:      path,
+		mounter:   mount.New(""),
+	}
+}
+
+// Run will only execute the cleanup logic if enough time has passed from the previous run, to not overload the IO of the node
+func (c Cleaner) Run(ctx context.Context) error {
+	tickerResetFunc := checkTicker()
+	if tickerResetFunc == nil {
+		return nil
+	}
+	defer tickerResetFunc()
+
+	return c.run(ctx)
+}
+
+// InstantRun will always execute the cleanup logic ignoring the time passed from previous run
+func (c Cleaner) InstantRun(ctx context.Context) error {
+	defer resetTickerAfterDelete()
+
+	return c.run(ctx)
+}
+
+func (c Cleaner) run(ctx context.Context) error {
+	fsState, err := c.getStateOfFilesystem()
+	if err != nil {
+		return err
+	}
+
+	c.removeDeprecatedMounts(fsState)
+
+	dks, err := metadata.GetRelevantDynaKubes(ctx, c.apiReader)
+	if err != nil {
+		log.Info("failed to list available dynakubes, skipping unused binaries cleanup")
+
+		return err
+	}
+
+	c.removeHostMounts(dks, fsState)
+
+	err = c.removeUnusedBinaries(dks, fsState)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c Cleaner) getStateOfFilesystem() (fsState fsState, err error) { //nolint:revive
```

Comment:
```suggestion
func (c Cleaner) getFilesystemState() (fsState fsState, err error) { //nolint:revive
```

Reply:
Reads better IMO, it is up to you if you agree.

Reply:
done 👍